[CORE] Quantized lm-head Framework

Qubitium opened the pull request:
MOTIVATION
* lm-head and be quantized with minimal loss in accuracy
* models like llama-3 have very large vocab sizes

SUMMARY:
* support quantized lm head (for models without tied word embeddings)
* update autogptq integration to enable quantized lm-head
* added tests for ~all models that can fit in our L4 CI to confirm this does not break anything

IMPLEMENTATION:
* refactored `VocabParallelEmbedding` to use `create_weights` to create the parameters
* refactored `ParallelLMHead` to use `apply()` to generate output
* plumb changes through all models
* plumb `quant_config` into `ParallelLMHead`
* updated AutoGPTQ integration to use `lm-head` if detected in config


FOLLOW UPS:
* support for fp8 and int8
* look into quantized embeddings


TEST MODEL: (quantized by auto-round and load tested with autogptq): 

- https://huggingface.co/LnL-AI/TinyLlama-1.1B-intermediate-step-1341k-3T-autoround-lm_head-symFalse

https://github.com/intel/auto-round/blob/8a3da144423322dfedb0b3fa702ae35d242496d8/docs/Meta-Llama-3-8B-Instruct-acc.md?plain=1#L3

| Metric           | **BF16** | w4g128 w/o lm-head | w4g128 with quantized lm-head |
| ---------------- | :------- |--------------------|---------------------------------|
| Avg.             | 0.6352   | 0.6312             | 0.6303                          |
| mmlu             | 0.6386   | 0.6306             | 0.6318                          |
| winogrande       | 0.7143   | 0.7238             | 0.7269                          |
| truthfulqa_mc1   | 0.3623   | 0.3537             | 0.3525                          |
| rte              | 0.6751   | 0.6859             | 0.6679                          |
| piqa             | 0.7867   | 0.7797             | 0.7802                          |
| openbookqa       | 0.3400   | 0.3300             | 0.3320                          |
| lambada_openai   | 0.7182   | 0.7200             | 0.7173                          |
| hellaswag        | 0.5769   | 0.5699             | 0.5701                          |
| boolq            | 0.8297   | 0.8309             | 0.8284                          |
| arc_easy         | 0.8152   | 0.8089             | 0.8106                          |
| arc_challenge    | 0.5299   | 0.5102             |  0.5154                              |




---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

Qubitium committed:
skip unknown param g_idx for gptq

Qubitium committed:
skip unknown param g_idx for gptq

Qubitium committed:
skip unknown param g_idx for gptq

Qubitium committed:
debug

Qubitium committed:
update

Qubitium committed:
update

Qubitium committed:
update

Qubitium committed:
test pass linear_method

Qubitium committed:
test pass linear_method

Qubitium committed:
pass linear_method

Qubitium committed:
req update

Commit:
fix issue: can't load quant lm_head weights

Commit:
call lm_head.linear_method.apply_weights()

Commit:
Merge branch 'main' into autogptq-lm-head

Commit:
use quant_config.get_quant_method()

Qubitium committed:
lm_head is VocabParallelEmbedding not LinearBase

Commit:
fix error: Overwriting existing tensor attribute: weight_loader

Qubitium committed:
lm_head is VocabParallelEmbedding not LinearBase

Qubitium committed:
merge

Qubitium committed:
merge

squash 1

Qubitium committed:
set separate_bias_add=True for unquant linear_method

Qubitium committed:
cleanup

Qubitium committed:
try to mimic original logic

Qubitium committed:
use linear_method only

Qubitium committed:
keep logic as close to main as possible

Qubitium committed:
wrong name

Qubitium committed:
move linear_method decoding to ParallelLmHead class

Commit:
fix error: UnquantizedLinearMethod.create_weights() missing 1 required positional argument: 'params_dtype'

Commit:
Fix get embedding error

Qubitium committed:
revert req.txt to main

Qubitium committed:
rename

Qubitium committed:
add QUANTIZED bool param to base LinearMethod so we can avoid all the isinstance calls

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/logits_processor.py:
```diff
@@ -36,7 +37,7 @@ def __init__(self,
     def forward(
         self,
-        embedding: torch.Tensor,
+        lm_head,
```
need a type

robertgshaw2-neuralmagic commented:
@Qubitium thanks for this!

We can get this merged very quickly as it was something I was planning to do this week

Qubitium committed:
wrong state

Qubitium committed:
pep

Qubitium committed:
override QUANTIZED property to True for all quant based linear methods

robertgshaw2-neuralmagic commented:
@Qubitium could you also support passing a quantization_config to the Embedding?

Qubitium committed:
comment TODO

Qubitium commented on vllm/model_executor/layers/vocab_parallel_embedding.py:
```diff
@@ -96,8 +122,12 @@ def forward(self, input_):
         else:
             masked_input = input_
-            # Get the embeddings.
+
+        # Get the embeddings.
+        # TODO: linear_method base class does not have embedding api
+        assert not self.linear_method.QUANTIZED
         output_parallel = F.embedding(masked_input, self.weight)
```
@robertgshaw2-neuralmagic  We are running into this issue right now. `linear_method` base has no `embedding` api so we can't just call `linear_method.embedding(masked_input)`. This can be adde to `UnquantizedLinearMethod` but we don't have the tool/skills to add this api to gptq/other qunatized methods. 

Should we leave this piece of code as-is with TODO comment?

robertgshaw2-neuralmagic replied:
We can leave as a TODO and address in another PR

Do you have any example models with:
- quantized lm head, no quantized embedding
- quantized lm head, quantized embedding

Qubitium committed:
pass quant_config to base class and init linear_method there

Qubitium committed:
ruff

Qubitium commented:
> @Qubitium could you also support passing a quantization_config to the Embedding?

Done.

Qubitium committed:
ruff

robertgshaw2-neuralmagic commented on requirements-cpu.txt:
```diff
@@ -3,4 +3,4 @@
 torch == 2.2.1+cpu
-triton >= 2.2.0  # FIXME(woosuk): This is a hack to avoid import error.
\ No newline at end of file
+triton >= 2.2.0  # FIXME(woosuk): This is a hack to avoid import error.
```
don't touch this file

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/vocab_parallel_embedding.py:
```diff
@@ -60,31 +67,57 @@ def __init__(self,
-                                      vocab_end_index]
-        param[:loaded_weight.shape[0]].data.copy_(loaded_weight)
+        if self.linear_method.QUANTIZED:
+            tp_rank = get_tensor_model_parallel_rank()
```
Please use the existing pattern we have for weight loading of quantized weights with `packed_dim`

The pattern is that:
- In `create_weights`, we set the `packed_dim` dictionary as a `parameter` member
- In `weight_loader`, we look check if `packed_dim` is set. If so we adjust

Qubitium replied:
@robertgshaw2-neuralmagic  We copied the closest matching loader code from `MergedColumnParallelLinear` and copied that here. It worked for in our testing but likely not the best or the correct solution.

Qubitium committed:
de-ruff

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/vocab_parallel_embedding.py:
```diff
@@ -60,31 +67,61 @@ def __init__(self,
+                                          self.embedding_dim,
+                                          [self.num_embeddings_per_partition],
+                                          self.embedding_dim,
+                                          self.num_embeddings_per_partition,
```
this should be total_embeddings

Qubitium replied:
@robertgshaw2-neuralmagic  output_size fixed to be total: `self.embedding_dim`

robertgshaw2-neuralmagic commented:
Looks like you're headed in the right direction.

LMK if you need anything.

Please add a couple models to the CI for testing

Qubitium committed:
revert newline

Qubitium committed:
fix bad class name QuantizeMethodBase to QuantizableMethodBase

Qubitium commented on vllm/model_executor/layers/quantization/base_config.py:
```diff
@@ -5,8 +5,9 @@
 
-class QuantizeMethodBase(ABC):
-    """Base class for different quantized methods."""
+class QuantizableMethodBase(ABC):
+    """Base class for different methods that may be quantized."""
+    QUANTIZED = False
```
@robertgshaw2-neuralmagic 

Renamed `QuantizeMethodBase` to more accurate `QuantizableMethodBase` since non-quantized methods also inherit this.

Qubitium committed:
Merge branch 'main' into autogptq-lm-head

Commit:
The output_size parameter of create_weights() should be total embeddings

Qubitium committed:
missed rename

Qubitium committed:
isort

Qubitium committed:
clean

Commit:
modify weight_loader()

Qubitium committed:
format

Qubitium committed:
short method name and apply to all models

Qubitium committed:
move method into correct util

Commit:
add test_quant_lm_head_layer()

Qubitium committed:
pass lm_head and not lm_head.weights

Qubitium committed:
Merge remote-tracking branch 'origin/autogptq-lm-head' into autogptq-lm-head

Commit:
add test_quant_lm_head_false()

Qubitium committed:
do not change opt/lm_head type

Qubitium committed:
Merge remote-tracking branch 'origin/autogptq-lm-head' into autogptq-lm-head

Qubitium committed:
pass test_logits_processor

Qubitium committed:
enforce_eager for more stable ci test

Qubitium committed:
revert model lm_head type to original

Qubitium committed:
standardize lm_head for ParallelLMHead

Qubitium committed:
ruff/import

Qubitium committed:
format

Qubitium committed:
Revert "standardize lm_head for ParallelLMHead"

This reverts commit 1e03ad8be0565605679eb2162151babc573821f2.

Qubitium committed:
fix oom in test

Qubitium committed:
fix starcoder2

Qubitium committed:
fix lora

Qubitium committed:
fix lora test

Qubitium committed:
pass quant_config to all ParallelLMHead

Qubitium commented:
@robertgshaw2-neuralmagic  All tests passing (except for out of disk Neuro-test). Going through all the models, not all model's lm_head is of type `ParallelLMHead`. Loading of gptq quantization is currently enabled for `ParallelLMHead` only.

Qubitium commented:
@robertgshaw2-neuralmagic  All tests passing (except for out of disk Neuro-test). Going through all the models, not all model's lm_head is of type `ParallelLMHead`. Loading of gptq quantization is currently enabled for `ParallelLMHead` only.

Add ci tests for lm_head quantized true/false. 

* [ ] TODO: merge the 2 test files into one.

Qubitium committed:
comment

Commit:
expand testing for refactor

Commit:
Merge remote-tracking branch 'upstream/main' into expand-testing

Commit:
expanded testing

Qubitium committed:
raise error if lm_head is quantized but layer is not ParallelLMHead

robertgshaw2-neuralmagic commented:
@Qubitium 

IMPORTANT - given this PR touches a lot of models, I expanded the testing strategy

https://github.com/Qubitium/vllm/pull/2

PR does three things:
- merged main into this branch (so I could have access to some new files I added for recent PR)
- expanded small model testing to cover bigger set of models
- added a "test_models_logprobs" strategy for testing mediumm ~7b models (we can ask sim

Commit:
merge lm_false and _true into one test file

Qubitium commented:
>>Loading of gptq quantization is currently enabled for `ParallelLMHead` only.
> 
> What happens if this case occurs? Do we fail loudly?


Raise error if quantized `lm_head` enabled and class is not `ParallelLMHead`  https://github.com/vllm-project/vllm/pull/4442/commits/fd1cf0e37c4c6ba0a1497c9de06ad3b45497e9a8

Added `is_lm_head` to `get_quant_method` since we must be told if layer is head or not. Many models have both `VocabParallelEmbedding` and `ParallelLMHead`. Without boolean, `get_quant_method` cannot tell them apart and catch this misconfiguration.

Commit:
formatting

Qubitium commented:
Modified ci test so that the test is more explicit in lm_head layer check.  https://github.com/vllm-project/vllm/pull/4442/commits/da0af51df35f5dc340806178d622b1f1aab6de6b

Qubitium committed:
format

Qubitium commented:
> > > Loading of gptq quantization is currently enabled for `ParallelLMHead` only.
> > 
> > 
> > What happens if this case occurs? Do we fail loudly?
> 
> Raise error if quantized `lm_head` enabled and class is not `ParallelLMHead` [fd1cf0e](https://github.com/vllm-project/vllm/commit/fd1cf0e37c4c6ba0a1497c9de06ad3b45497e9a8)
> 
> Added `is_lm_head` to `get_quant_method` since we must be told if layer is head or not. Many models have both `VocabParallelEmbedding` and `ParallelLMHead`. Without boolean, `get_quant_method` cannot tell them apart and catch this misconfiguration.

Hold on. Some models have `lm_head` as VocabParallelEmbedding. So lm_head can be both VocabParallelEmbedding and ParallelLMHead. The commit as stands if not good.

Qubitium committed:
Merge pull request #2 from neuralmagic/expand-testing

expand testing for refactor

Qubitium committed:
Revert "raise error if lm_head is quantized but layer is not ParallelLMHead"

This reverts commit fd1cf0e37c4c6ba0a1497c9de06ad3b45497e9a8.

Qubitium committed:
fix test

Qubitium commented:
> @Qubitium
> 
> IMPORTANT - given this PR touches a lot of models, I expanded the testing strategy
> 
> [Qubitium#2](https://github.com/Qubitium/vllm/pull/2)
> 
> PR does three things:
> 
> * merged main into this branch (so I could have access to some new files I added for recent PR)
> * expanded small model testing to cover bigger set of models
> * added a "test_models_logprobs" strategy for testing mediumm ~7b models (we can ask sim

@robertgshaw2-neuralmagic Merged but tests are failing and appears unrelated to the changes made in this PR. Can you check?  I am not familiar with the src/causation of the errors. Thanks.  hf_runner missing `max_model_len`
https://buildkite.com/vllm/ci/builds/6110#018f2d04-809c-43bd-99b0-4e7aa6165a73

Qubitium commented:
We are currently adding opt-2.7b test. If successful then we can remove ParallelLmHead limitation. Many models' lm_head is the same embedding and not separate module. So we will cover all cases with llama and opt-2.7b lm_head quantize test.

Commit:
Skip loading of weights of lm_head layer.
There will be four weights(lm_head.g_idx, lm_head.qweight, lm.qzeros, lm.scales) after gptq quantizes the lm_head layer.

Commit:
test for opt-125M

robertgshaw2-neuralmagic commented:
> > @Qubitium
> > IMPORTANT - given this PR touches a lot of models, I expanded the testing strategy
> > [Qubitium#2](https://github.com/Qubitium/vllm/pull/2)
> > PR does three things:
> > 
> > * merged main into this branch (so I could have access to some new files I added for recent PR)
> > * expanded small model testing to cover bigger set of models
> > * added a "test_models_logprobs" strategy for testing mediumm ~7b models (we can ask sim
> 
> @robertgshaw2-neuralmagic Merged but tests are failing and appears unrelated to the changes made in this PR. Can you check? I am not familiar with the src/causation of the errors. Thanks. hf_runner missing `max_model_len` https://buildkite.com/vllm/ci/builds/6110#018f2d04-809c-43bd-99b0-4e7aa6165a73

Will fix

Qubitium committed:
rename VocabParallelEmbedding to ParallelVocabEmbedding for consistency

Qubitium committed:
Merge remote-tracking branch 'origin/autogptq-lm-head' into autogptq-lm-head

Qubitium commented:
Running a big issue loading OPT quantized `lm_head` due to the fact for OPT (and other models) `lm_head` just a soft-linked to embeddings (not a unique lm_head layer like llama).  Checking with intel/auto-round team to see if this is a potential issue that should be addressed at the quantization stage or a remapping (here in vlllm) that we need to do. I just don't know the correct answer yet. 

Once this is addressed, I believe limits to `ParallelLmHead` can be fully unlocked and we just need to check for `ParallelVocabEmbedding`

https://github.com/intel/auto-round/issues/100

Qubitium committed:
fix opt not loading quantized lm_head

robertgshaw2-redhat committed:
Merge branch 'main' into autogptq-lm-head

robertgshaw2-neuralmagic commented:
> Running into a big issue loading OPT quantized `lm_head` due to the fact for OPT (and other models) `lm_head` just a soft-linked in code to embeddings (not a unique lm_head layer like llama). But the thing is OPT model has both lm_head and embedding in original weights except they are the same tensor shape/size and values.
> 
> Checking with intel/auto-round team to see if this is a potential issue that should be addressed at the quantization stage or a remapping (here in vlllm) that we need to do. I just don't know the correct answer yet. This could be a bug in vllm OPT code in which assumptions are made that lm_head is always the same as embedding and should be ignored on load. This may be true for pre-quant but post-quant it may not be.
> 
> Once this is addressed, I believe limits to `ParallelLmHead` can be fully unlocked and we just need to check for `ParallelVocabEmbedding`
> 
> Asking Intel/auto-round devs for clarification/insight: [intel/auto-round#100](https://github.com/intel/auto-round/issues/100)
> 
> Change PR to draft-mode until this issue is resolved.

Let's handle this in a follow up PR since the scope of this is already big

Commit:
updated test, opt model is failing

Qubitium committed:
fix opt and disable marlin when lm_head quant is enabled. TODO find out why marlin load is broken

Qubitium committed:
Merge remote-tracking branch 'origin/autogptq-lm-head' into autogptq-lm-head

# Conflicts:
#	tests/quantization/test_quant_lm_head.py

Qubitium commented:
I overwrote your test changes in commit https://github.com/vllm-project/vllm/pull/4442/commits/2f63a723726bd4cd06a06c21c51fdaf360fbec8a

Will re-merge with your changes later. Fixed opt compat with lm_head for now.

New problem Marlin kernel loading with lm_head is broken. For now I disabled Marlin auto-upconvert when `lm_head` is detected/True.

* [ ] Fix Marlin `lm_head`

Qubitium committed:
remove debug prints

Qubitium committed:
isort

robertgshaw2-neuralmagic commented:
@Qubitium no worries - to make it easier for us to both work on it, im going to move the model testing refactor to another PR

Qubitium committed:
add non-quantized opt-125m to test to ensure that quant code did not break non-quant

robertgshaw2-neuralmagic commented:
@Qubitium I'm going to merge this first

https://github.com/vllm-project/vllm/pull/4510

Qubitium committed:
1. move lm_head_quantized proper into base QuantizationConfig class.

2. optimize opt vram usage by keeping only one copy of lm_head vs embed_tokens when possible

Qubitium committed:
format

Qubitium commented on vllm/model_executor/models/opt.py:
```diff
@@ -351,3 +351,9 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
                                         default_weight_loader)
                 weight_loader(param, loaded_weight)
+
+        # memory optimization: if lm_head quantization is not enabled lm_head
+        # is same as embed_tokens. keep only one copy
+        if (self.quant_config is None
+                or not self.quant_config.lm_head_quantized):
+            self.lm_head = self.model.decoder.embed_tokens
```
@robertgshaw2-neuralmagic  I moved lm_head_quantized property into base `QuantizationConfig` so that this code here doesn't have nasty isinstance checks for all quant configs that suppor t`lm_head` quants. Most likely other quant models/code will also need this check.

Qubitium commented:
@robertgshaw2-neuralmagic  I do not plan to make any more changes unless pending CI build shows something broken.  https://buildkite.com/vllm/ci/builds/6198 

Feel free to add/mod. 

Changes and notes:

* Moved `lm_head_quantized` property from GPTQConfig to bas QuantizationConfig.
* Fixed OPT compat 
* OPT will on `load_weights` end will delete `model.lm_head` (softlink to `embed_tokens`  if model is not quantized or quantized with lm_head false as not to duplicate memory as in this case, the weights are the same. 

On last point, it may be a good idea to add an api (future PR) to model so there is an equivalent `on_load_weights_end` method that does model weight cleanups/logic post load.

Qubitium commented:
All the relevant tests are passing. Of note that the code changes for OPT works but is actually not efficient. There is negative advantage to have an OPT model enable quantization of `lm_head` as it  would, at the moment, disable the sharing of embed_tokens and load 2 different layers causing he overall OPT memory footprint to increase. For llama models where lm_head are separate, the memory saving is significant.

Qubitium committed:
fix marlin loading of lm_head. remove TODO

Commit:
add comments for weight_loader

Qubitium committed:
Replace BaseConfig.is_lm_head_quantized property with BaseConfig.is_lm_head_quantized() interface method

Qubitium committed:
Merge remote-tracking branch 'origin/autogptq-lm-head' into autogptq-lm-head

Qubitium commented:
To align with `BaseConfig` inheritance by interface and not property vars, refractored `BaseConfig.lm_head_quantized` variable into `BaseoConfig.is_lm_head_quantized()` interface. https://github.com/vllm-project/vllm/pull/4442/commits/9981620f93d97b9fa50bfb0eb8fcabc20cd18df8

Qubitium committed:
implement BaseConfig.is_lm_head_quantized() for all non-gptq quant configs

Qubitium committed:
fix lora logits bias computed twice

Qubitium committed:
format

robertgshaw2-neuralmagic commented:
Going to pick this back up this weekend

Commit:
merged - removed quantized lm head for tied embeddings

robertgshaw2-neuralmagic commented:
@Qubitium - I apologize for not getting this merged before. I am picking up the PR now

Commit:
format

Commit:
format passing

Commit:
fix all parallel-vocab-embedding --> vocab-parallel-embedding

Commit:
quantizable-method-base --> quantize-method-base

Commit:
spurious change

Commit:
cleanup AQLM

Commit:
missed nit

Commit:
fix up awq

Commit:
fix up fp8

Commit:
nit

Commit:
cleanup gptq marlin

Commit:
cleanup more

Commit:
nit

Commit:
remove QUANTIZED

Commit:
if -- elif

Commit:
removed

Commit:
add kwarg

Commit:
cleanup

Commit:
more cleanup

Commit:
nit

Commit:
use linear_method.create_weights

Commit:
fix spurious changes

Commit:
fix typo

Commit:
remove another spurious change

Commit:
format

Commit:
format fix

Commit:
passing linting

Commit:
fix more changes

Commit:
spurious newline

Commit:
spurious change in qwen

Commit:
first end to end run of llama working!

Commit:
format

Commit:
cleanup

Commit:
Merge branch 'main' into autogptq-lm-head

Commit:
format

Commit:
fix opt

Commit:
fix gpt2

Commit:
fix arctic

Commit:
fix gptbigcode

Commit:
actually fix gptbigcode

Commit:
commandr

Commit:
fix deepseekv2

Commit:
fix falcon

Commit:
fix gemma2

Commit:
fix jais

Commit:
fix mpt

Commit:
fix phi3 small

Commit:
fix vision models

Commit:
all models should run

Commit:
formatted

robertgshaw2-neuralmagic commented on vllm/model_executor/models/commandr.py:
```diff
@@ -362,12 +362,12 @@ def compute_logits(self, hidden_states: torch.Tensor,
         is_not_lora = hasattr(self.model.embed_tokens, 'weight')
         if is_not_lora:
-            embedding_weights = self.model.embed_tokens.weight
+            logits = self.logits_processor(self.model.embed_tokens,
```
@Yard1 - do you know why Command-R has this lora special case?

Commit:
cleanup lora test changes

Commit:
cleanup tests

Commit:
nit

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/vocab_parallel_embedding.py:
```diff
@@ -288,10 +301,32 @@ def get_sharded_to_full_mapping(self) -> Optional[List[int]]:
-        assert loaded_weight.shape[parallel_dim] == self.org_vocab_size
-        loaded_weight = loaded_weight[self.shard_indices.org_vocab_start_index:
-                                      self.shard_indices.org_vocab_end_index]
+        output_dim = getattr(param, "output_dim", None)
```
Note: very similar logic to `ColumnParallelLinear`

Commit:
remove llama debug

Commit:
fix spurious change in qwen2

Commit:
formats again

robertgshaw2-neuralmagic commented:
cc 

- @Yard1 for `VocabParallelEmbedding` weight_loading --> anything to be aware of for LoRA?
- @comaniac for quantization changes

robertgshaw2-redhat reviewed (commented):
I contributed a significant amount of code to this PR, so I need another review

Commit:
re-enabled tests

Commit:
format

Commit:
skip fp16 tests in non cuda automation

Commit:
fix mlp-speculator

Commit:
fix spec decode tests

robertgshaw2-neuralmagic commented:
cc @njhill - had to make a couple changes to `MLPSpeculator`, which uses the `lm_head`

When I did this, the e2e tests for `MLPSpeculator` failed because of logits divergence in the middle of a sequence at a random batch index. When I tweaked the precision the test runs at to fp32, the issue went away, suggesting to me that it is numerics.

This PR changes the `logit_processor` from running `torch.matmul(hidden, embedding.t()` we run with `F.Linear()` 

Do you think this is okay? Or emblematic of an issue?

Commit:
passing mlp correctness with precision of float32

robertgshaw2-neuralmagic commented on tests/spec_decode/e2e/test_mlp_correctness.py:
```diff
@@ -34,7 +34,7 @@
 
 # precision
-PRECISION = "float16"
+PRECISION = "float32"
```
@njhill @cadedaniel 

Would you be okay with this change?

robertgshaw2-neuralmagic replied:
Welp, now we are OOM in the CI

robertgshaw2-neuralmagic replied:
Will dive deeper

robertgshaw2-neuralmagic replied:
Nick is making another model FYI

cadedaniel replied:
thanks @njhill !

Commit:
Merge branch 'upstream-main' into autogptq-lm-head

Commit:
Merge branch 'upstream-main' into autogptq-lm-head

Commit:
format

robertgshaw2-redhat committed:
Merge branch 'main' into autogptq-lm-head

comaniac commented on vllm/model_executor/layers/quantization/base_config.py:
```diff
@@ -87,6 +87,15 @@ def get_from_keys(config: Dict[str, Any], keys: List[str]) -> Any:
                          "quantization config.")
 
+    @staticmethod
+    def get_from_keys_optional(config: Dict[str, Any], keys: List[str],
```
Suggest naming: `get_from_keys_or`

Commit:
Address cody's comments

Commit:
Merge branch 'autogptq-lm-head' of https://github.com/Qubitium/vllm into autogptq-lm-head

Commit:
format

Commit:
updated names

Commit:
format

robertgshaw2-redhat committed:
Merge branch 'main' into autogptq-lm-head

robertgshaw2-neuralmagic commented:
@comaniac ready for re-review

(ignore spec decode test failure @njhill making a new model that avoids OOM at fp32)

comaniac commented on tests/conftest.py:
```diff
@@ -437,7 +437,7 @@ def __init__(
             model=model_name,
             tokenizer=tokenizer_name,
-            trust_remote_code=True,
+            trust_remote_code="falcon" not in model_name,
```
Why we have to re-download Falcon?

robertgshaw2-neuralmagic replied:
`trust_remote_code` did not work for falcon, not sure why

comaniac commented on tests/models/test_models_logprobs.py:
```diff
@@ -0,0 +1,90 @@
+
+MAX_MODEL_LEN = 1024
+
+MODELS = [
```
How long does it take to run all models listed here? Can some of them be removed to reduce the CI time?

robertgshaw2-neuralmagic replied:
I can just remove them. I just wanted to prove the accuracy was right

comaniac reviewed (approved):
Overall LGTM

Commit:
remove too long tests

Commit:
cleanup tests

Commit:
updated

njhill commented:
> 
> * @Yard1 for `VocabParallelEmbedding` weight_loading --> anything to be aware of for LoRA?


@robertgshaw2-neuralmagic @Yard1 it looks like there was some impact from this ... not sure if it actually exposed a latent bug where the lm_head for gpt_bigcode (and similar) was not previously adaptable: https://github.com/vllm-project/vllm/issues/6314

robertgshaw2-neuralmagic commented:
> > * @Yard1 for `VocabParallelEmbedding` weight_loading --> anything to be aware of for LoRA?
> 
> @robertgshaw2-neuralmagic @Yard1 it looks like there was some impact from this ... not sure if it actually exposed a latent bug where the lm_head for gpt_bigcode (and similar) was not previously adaptable: #6314

@njhill do you know if this was working before?

RanchiZhao commented:
hi！is 'look into quantized embeddings' available now? looking forward to this!

RanchiZhao commented:
hi！is 'look into quantized embeddings' available now? looking forward to this!
